ADBDEV-4726-67 Fix null-check for par_prule

dnskvlnk opened the pull request:
Add assertion for par_prule

wack_pid_relname dereferences par_prule before null-check making segfault
possible. This patch adds an assertion

Commit:
Fix null-check for par_prule

wack_pid_relname dereferences par_prule before null-check making segfault
possible. This patch adds moves dereferencing below null-check

RekGRpth commented on src/backend/commands/tablecmds.c:
```diff
@@ -16058,10 +16058,11 @@ wack_pid_relname(AlterPartitionId 		 *pid,
 
-		*plrelname = par_prule->relname;
-
 		if (par_prule && par_prule->topRule && par_prule->topRule->children)
```
I think you should have just replaced the `par_prule &&` condition with an assert.

dnskvlnk replied:
Wait, how that would help with assert works on debug builds only?

RekGRpth replied:
You should remove condition and add assert

dnskvlnk replied:
Can you explain why I should add assert if the analyzer doesn't see it?

RekGRpth replied:
for testing in debug build on CI

dnskvlnk replied:
But that wouldn't fix the problem with the analyzer

RekGRpth replied:
problem with the analyzer is fixed by removing condition

dnskvlnk replied:
But you told earlier that using assert is not a solution. When I should use Insist and when I should use assert?

RekGRpth replied:
in this case you may remove condition and add assert
in another case you may add insist, i.e. add new condition

dnskvlnk replied:
Done

Commit:
Fix null-check for par_prule

wack_pid_relname dereferences par_prule before null-check making segfault
possible. This patch adds an Assert

RekGRpth commented:
update PR description

Commit:
Add assertion for par_prule

wack_pid_relname dereferences par_prule before null-check making segfault
possible. This patch adds an assertion

RekGRpth commented:
update PR description too

bandetto commented on src/backend/commands/tablecmds.c:
```diff
@@ -16058,9 +16058,11 @@ wack_pid_relname(AlterPartitionId 		 *pid,
 
-		if (par_prule && par_prule->topRule && par_prule->topRule->children)
+		*plrelname = par_prule->relname;
+		
```
```suggestion

```
Remove trailing whitespace

dnskvlnk replied:
It was there before the patch, so I don't think it should be removed

bandetto replied:
I don't see it in the diff. It also breaks this PR's diff a little. Without whitespace your changes look like:
```diff
+               Assert(par_prule);
+
                *plrelname = par_prule->relname;
 
-               if (par_prule && par_prule->topRule && par_prule->topRule->children)
+               if (par_prule->topRule && par_prule->topRule->children)
```
What it looks like now:
```diff
-               *plrelname = par_prule->relname;
+               Assert(par_prule);
 
-               if (par_prule && par_prule->topRule && par_prule->topRule->children)
+               *plrelname = par_prule->relname;
+               
+               if (par_prule->topRule && par_prule->topRule->children)
```

dnskvlnk replied:
I see, you mean extra tab. Done

RekGRpth commented on src/backend/commands/tablecmds.c:
```diff
@@ -16058,9 +16058,11 @@ wack_pid_relname(AlterPartitionId 		 *pid,
 
-		*plrelname = par_prule->relname;
+		Assert(par_prule);
 
```
remove this whitespace between assert and dereference

dnskvlnk replied:
Is that _really_ needed?

dnskvlnk replied:
Do you mean extra new line or am I missing anything?

RekGRpth replied:
yes extra new line between assert and dereference

dnskvlnk replied:
Done

RekGRpth replied:
No, I meant between `Assert(par_prule);` and `*plrelname = par_prule->relname;`

dnskvlnk replied:
Assertion should be at the place where a variable is initialized or assigned, postgres also uses this style

Commit:
Remove extra tab

Commit:
Remove extra line
